server/src/main/java/org/opensearch/gateway/remote/IndexMetadataUploadListener.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.core.action.ActionListener;
+
+import java.util.List;
+
+/**
+ * Hook for running code that needs to be executed before the upload of index metadata. Here we have introduced a hook
+ * for index creation (also triggerred after enabling the remote cluster statement for the first time). The listener
+ * is intended to be run in parallel and async with the index metadata upload.
+ *
+ * @opensearch.internal
+ */
+public interface IndexMetadataUploadListener {
+
+    /**
+     * Runs before the new index upload of index metadata (or first time upload). The caller is expected to trigger
+     * onSuccess or onFailure of the {@code ActionListener}.
+     *
+     * @param indexMetadataList list of index metadata of new indexes (or first time index metadata upload).
+     * @param actionListener    listener to be invoked on success or failure.
+     */
+    void beforeNewIndexUpload(List<IndexMetadata> indexMetadataList, ActionListener<Void> actionListener);
+
+    String getThreadpoolName();
```

Comment:
> The individual executions of beforeNewIndexUpload  should decide how they wish to run

We intentionally don't want this to happen to avoid the listener code blocking the primary execution. It can also lead to cases where the timeout of the original index metadata upload does not happen since the listener is running the code as per it's own wish. We are making sure that the listener runs without blocking the remote cluster state service code and at the same time the listener provides using which threadpool it wants to execute the code asynchronously.

Reply:
Making the listener abstract and moving the code for forking the thread within the listener itself.